### PR TITLE
Add salmon processing pipeline

### PR DESCRIPTION
**Pull request by ajlee21**
This PR add a notebook `sra_experiment/1_download_process_data.ipynb` which takes a small sample of RNA-seq data downloaded from SRA and processes them using Salmon.

This notebook is pilot testing this processing step, which will be a part of this pipeline to process more RNA-seq samples for Pseudomonas aeruginosa and examine the relationship between core and accessory genes (see the ReadMe). This pipeline will also be used on Dartmouth's Discovery server to help with a project with a collaborator.

### CODE REVIEW

File: sra_experiment/1_download_process_data.ipynb [hunk 1]
```diff
@@ -0,0 +1,12781 @@
...
+      "\u001b[00m\u001b[00m[2020-07-08 15:49:21.816] [puff::index::jointLog] [info] finished populating pos vector\n",
+      "\u001b[00m\u001b[00m[2020-07-08 15:49:21.816] [puff::index::jointLog] [info] writing index components\n",
+      "\u001b[00m\u001b[00m[2020-07-08 15:49:21.828] [puff::index::jointLog] [info] finished writing dense pufferfish index\n",
+      "\u001b[00m[2020-07-08 15:49:21.829] [jLog] [info] done building index\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Get PAO1 index\n",
+    "! salmon index -t ~/Documents/Data/Core_accessory/Pseudomonas_aeruginosa_PAO1_107.ffn.gz -i ~/Documents/Data/Core_accessory/pao1_index"
```
Comment (arielah) [inline sra_experiment/1_download_process_data.ipynb:359]: This is sooo much output. I'm not a jupyter expert, but is there a way to write the stdout to a txt file instead of having it be written into the document? I imagine you won't need to reference it as much as the notebook itself, and that would save you like 10,000 lines.

File: sra_experiment/1_download_process_data.ipynb [hunk 2]
```diff
@@ -0,0 +1,12781 @@
...
+      "| Loading reference accumulative lengths | Time = 35.581 us\n",
+      "-----------------------------------------\n",
+      "[2020-07-08 15:53:46.845] [jointLog] [info] done\n",
+      "[2020-07-08 15:53:46.845] [jointLog] [info] Index contained 5,685 targets\n",
+      "[2020-07-08 15:53:46.846] [jointLog] [info] Number of decoys : 0\n",
+      "\n",
+      "\n",
+      "\n",
+      "\n",
+      "[2020-07-08 15:53:47.759] [jointLog] [info] Automatically detected mos
```
Comment (arielah) [inline sra_experiment/1_download_process_data.ipynb:11852]: I've only done alevin, not salmon, but it's weird that quantification and converting to a matrix happen separately. In alevin, you only need two commands to run a sample "salmon index" and "salmon alevin", and it returns a cell-by-gene matrix for you. Not sure if there's an analogous function for bulk data, but it could save some space and effort.

Comment (arielah) [inline sra_experiment/1_download_process_data.ipynb:12463]: As an outsider, I would benefit from a little two sentence walkthrough of these graphs. What do blue and red represent? What do these results say about the quality of your data?

File: sra_experiment/1_download_process_data.ipynb [hunk 3]
```diff
@@ -0,0 +1,12781 @@
...
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Extract tar.gz file \n",
+    "if not os.path.exists(\"sratoolkit.current-centos_linux64.tar.gz\"):\n",
+    "    ! tar -xzf sratoolkit.current-centos_linux64.tar.gz\n",
+    "\n",
+    "# add binaries to path using export path or editing ~/.bashrc file\n",
+    "! export PATH=$PATH:/home/alexandra/Documents/Data/Core_accessory/sratoolkit.2.10.7-centos_linux64/bin\n",
```
[the remark this answers is not in the document]
Comment (rando2) [reply sra_experiment/1_download_process_data.ipynb:82]: I actually cannot get this step to work! Not sure why Jupyter notebook isn't able to edit my path. I can keep looking into this but figured I'd provide this first set of feedback for now!

File: sra_experiment/1_download_process_data.ipynb (hunk 2 above)
Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb:11852]: Normally people use `salmon quant` to quantify transcripts and then port those quantifications to `tximport` to map transcripts to genes, which then outputs the data as sample x gene. But since prokaryotes are mostly 1-1 between genes and transcripts, I am only using this salmon step to quantify reads per gene and thus have to do the formatting myself.

Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb:12463]: Good point. This was a quick validation to check that i processed the data correctly. So I have two different strains and I expected to see that they separate, which they do

File: sra_experiment/1_download_process_data.ipynb (hunk 1 above)
Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb:359]: I very much agree. Looks like its a feature for downstream analysis to print to stdout: https://github.com/COMBINE-lab/salmon/issues/242

File: environment.yml [hunk 4]
```diff
@@ -7,16 +7,20 @@ dependencies:
...
 - conda-forge::python=3.7
 - conda-forge::pandas=0.24.2
 - conda-forge::seaborn=0.10
 - conda-forge::scipy=1.1.0
 - conda-forge::plotnine=0.5.1
+- conda-forge::scikit-learn
+- conda-forge::umap-learn
 - conda-forge::ipykernel=5.1.1
 - conda-forge::nb_conda_kernels
 - conda-forge::gxx_linux-64
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply environment.yml:20]: Thanks for checking the env setup. I will add this to the yml file.

File: README.md [hunk 5]
```diff
@@ -27,4 +27,8 @@ To build and activate this environment run:
 conda env create -f environment.yml
 
 conda activate core_acc
+```
+Also need to pip install modules from this repository:
+```bash
+pip install .
 ```
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply README.md:34]: This repo is still in the piloting stage to test trends in the data, so I haven't sat down to write out the steps for how to run the analysis just yet. But this is a very good point. I will try to add some of this now and add more in the future

File: sra_experiment/1_download_process_data.ipynb [hunk 6]
```diff
@@ -0,0 +1,12781 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Extract tar.gz file \n",
+    "if not os.path.exists(\"sratoolkit.current-centos_linux64.tar.gz\"):\n",
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb:78]: good catch!

**Comment by ajlee21** [comment]
Thank you so much for trying to re-run this analysis...I have added a paths.py file that can be modified to make this more user-friendly hopefully and help to run things..keep me posted.

File: sra_experiment/1_download_process_data.ipynb [hunk 7]
```diff
@@ -0,0 +1,12880 @@
...
+      "\n",
+      "\n",
+      "\n",
+      "\n",
+      "\n",
+      "\n",
+      "\n",
+      "[2020-07-10 17:59:59.702] [jointLog] [info] Computed 7,767 rich equivalence classes for further processing\n",
+      "[2020-07-10 17:59:59.702] [jointLog] [info] Counted 1,149,139 total reads in the equivalence classes \n",
+      "[2020-07-10 17:59:59.704] [jointLog] [i
```
Comment (rando2) [inline sra_experiment/1_download_process_data.ipynb:12547]:
```suggestion
    "If we processed the data correctly, we'd expect that genes that are PAO1-specific (ie. those PAO1 genes that do not have a PA14 homolog) typically have very low (0) expression in PA14 samples. We do see this - blue has many genes that have 0 mean gene expression, while the red does not. "
```

Comment (rando2) [inline sra_experiment/1_download_process_data.ipynb:12639]: I may be overly intense about this but would it make sense to set the random seed as a variable at the beginning?

Comment (rando2) [reply sra_experiment/1_download_process_data.ipynb:12547]: This is not quite working, but basically just that sentence is a bit hard to follow so I suggested a way to rephrase it a little.

Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb:12547]: Thanks I have gone back to reword this so hopefully its more clear

File: sra_experiment/nbconverted/1_download_process_data.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+# In[3]:
+
+
+# Extract tar.gz file 
+if not os.path.exists("sratoolkit.current-centos_linux64.tar.gz"):
+    get_ipython().system(' tar -xzf sratoolkit.current-centos_linux64.tar.gz')
+
+# add binaries to path using export path or editing ~/.bashrc file
+get_ipython().system(' export PATH=$PATH:/home/alexandra/Documents/Data/Core_accessory/sratoolkit.2.10.7-centos_linux64/bin')
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply sra_experiment/nbconverted/1_download_process_data.py]: I will add a comment about this in the readme

File: sra_experiment/1_download_process_data.ipynb (hunk 7 above)
Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb:12639]: always a good idea! thank you

**Comment by ajlee21** [comment]
@rando2 thanks for taking a second look. I've update the files based on your nice suggestions. Please let me know if you think its ok to merge or if you have other suggestions

**Comment by rando2** [comment]
> @rando2 thanks for taking a second look. I've update the files based on your nice suggestions. Please let me know if you think its ok to merge or if you have other suggestions

Hi @ajlee21, it looks good to me!

File: sra_experiment/1_download_process_data.ipynb (hunk 7 above)
Comment (rando2) [reply sra_experiment/1_download_process_data.ipynb]: Sorry if my comment seemed rude -- I meant something was going wrong with the suggest interface and I wasn't actually suggesting you change your log output but rather the text!

Comment (ajlee21) [reply sra_experiment/1_download_process_data.ipynb]: Not at all rude! I knew what you meant no worries
